Upload backup volume to image

wwba opened the pull request:
This patch implements creating image from backup of volume, using
ceph backend clone or a local tempfile uploading.

Signed-off-by: Xiaojun Liao <xiaojunliao85@gmail.com>

Commit:
Fix properties extracting from image with glance api v2

When a volume is created using the glance v1 API, the non based
properties places in 'properties' dict, and nova is expected
to find them here. Otherwise api's v2 return custom image properties
as base properties, so nova can't find them in volume 'image
properties'.

This change add new glance call to get image schema, it allows
extract custom properties from image.

Closes-Bug: 1323660

Change-Id: Ib54bb6759b27334294fb2c6d2c0bfe4eae3d0920
(cherry picked from commit f20643ca44e139e3384b9dea638b54c2235b1169)

Conflicts:
	cinder/tests/image/test_glance.py

Commit:
Fix always false condition in glance wrapper

If statement's condition in GlanceClientWrapper's method 'call'
always false, because 'version' key passed through kwargs is
string, and variable chacked in 'call' is None or int.

Change-Id: Ief1e8d2c41aaa774666bc59d63cdea68f80f308a
Closes-Bug: #1439275
(cherry picked from commit 486da4305d607b32b03709c6b0f26430776a37cf)

Commit:
fix cinder ceph backup driver padding error

While restore an extended source, ceph backup driver will pad
0 to the unused space if restore_discard_excess_bytes is True.
The original code use total length of the unused space to fill
each chunk. Change to use chunk size to fill each chunk.

Change-Id: I9658d31a55e5024f5f6d5ccf00939edd668ea451
Signed-off-by: Liu Qing <liuqing@chinac.com>
Closes-Bug: 1595014

Conflicts:
	cinder/tests/test_backup_ceph.py

wwba committed:
Upload backup volume to image

Fixes: Feature #10587

This patch implements creating image from backup of volume, using
ceph backend clone or a local tempfile uploading.

Signed-off-by: Xiaojun Liao <xiaojunliao85@gmail.com>

zhaochao commented on cinder/api/contrib/backups.py:
```diff
@@ -30,6 +30,9 @@
 from cinder.openstack.common import log as logging
 from cinder import utils
 
+from oslo import messaging
```
这里没有必要，在 api 里不应该直接调用 oslo.messaging 里的代码。

后面捕获 oslo.messaging.RemoteError 也是不必要的。

wwba replied:
是不不要的. bakcup.py中调下去的接口都没有处理远端的异常. 最开始参参考volume_actions.py中_volume_upload_image的处理,backup.py这边处理应该不一样.

zhaochao commented on cinder/api/contrib/backups.py:
```diff
@@ -267,6 +270,74 @@ def create(self, req, body):
+            raise webob.exc.HTTPBadRequest(explanation=error.msg)
+        except ValueError as error:
+            raise webob.exc.HTTPBadRequest(explanation=unicode(error))
+        except messaging.RemoteError as error:
```
在 rpcapi 中用的是 cast 方式，这里不会返回异常。

而且更底层的异常也是在 backup/api.py 、 backup/manager.py 、backup/drivers 逐级处理的，到 cinder/api 这里应该至少是明确的异常，再根据异常返回具体的 http 响应。

所以不用专门去捕获 oslo.messaging 的异常。

zhaochao commented on cinder/api/contrib/backups.py:
```diff
@@ -267,6 +270,74 @@ def create(self, req, body):
+            msg = "%(err_type)s: %(err_msg)s" % {'err_type': error.exc_type,
+                                                 'err_msg': error.value}
+            raise webob.exc.HTTPBadRequest(explanation=msg)
+        except Exception as error:
```
这里和上面的 olso.messaging.RemoteError 类似，在 cinder/api 这里至少异常已经是明确的。

zhaochao commented:
在 Redmine 上提个 issue 吧，以后 pull request 尽量有对应的 issue 。

然后 commit message 里也加上（和其它内容之间空上一行）：
Fixes: xxx

xxx 就是对应 Redmine issue 。

zhaochao commented on cinder/backup/api.py:
```diff
@@ -202,6 +207,60 @@ def _consumed(resource_name):
+    def copy_backup_to_image(self, context, backup, metadata, force):
+        """Create a new image from the specified backup."""
+
+        if backup['status'] not in ['available',]:
```
多了 1 个 “,” ？

另外，force 这个参数，是这里要起作用的，不然 force 这参数就没有意义了。

wwba replied:
目前只支持'avaliable'一种状态,为了添加其他状态的同时允许客户端进行'force'参数设置,这里'force'参数没有去掉. 这里只有一种状态所以'force'参数用不上.

zhaochao replied:
只支持 `available` 的话，逗号就多余了。

不允许 force 的话，应该去掉这个参数的支持。

不过支持 force 的也没问题，但是 force 也要限定在 `available` 和 `restoring` 这两种状态，别的状态，force 也不应该允许。

zhaochao commented on cinder/backup/api.py:
```diff
@@ -202,6 +207,60 @@ def _consumed(resource_name):
+            recv_metadata = self.image_service.create(context,
+                         metadata)
+        except Exception:
+            with excutils.save_and_reraise_exception():
```
异常的问题参看前面几个评论。

如果不在 cinder/api 里对所有 Exception 作捕获的话，这里就需要 reraise 一个明确的异常，之后 cinder/api 那里再进行处理。

zhaochao commented on cinder/backup/api.py:
```diff
@@ -202,6 +207,60 @@ def _consumed(resource_name):
+            response['protected'] = recv_metadata.get('protected')
+        if 'is_public' in recv_metadata:
+            response['is_public'] = recv_metadata.get('is_public')
+        LOG.info(_("backup %(backup_id)s to image %(image_id)s completed successfully."),
```
因为是异步进行的，这个日志信息是不对的，这里只是开始上传。

另外 1 行太长了，需要换行。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -88,7 +94,13 @@
                 help='If True, always discard excess bytes when restoring '
-                     'volumes i.e. pad with zeroes.')
+                     'volumes i.e. pad with zeroes.'),
+    cfg.StrOpt('backup_cluster_name',
```
这个参数没必要，--conf 已经确定了 cluster ，已经指定了 backup_ceph_conf ，就没必要再加上 --cluster 。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1147,144 @@ def _restore_metadata(self, backup, volume_id):
             LOG.error(msg)
             raise exception.BackupOperationError(msg)
 
+    def _is_non_recoverable(self, err, non_recoverable_list):
```
_is_non_recoverable 和 _try_execute 是 cinder/volume/driver.py 里的代码？不应该再复制一遍。

如果确定有必要的话，应该把这部份抽出来做成公共的接口，让 cinder/volume/drver.py 和这里都可以使用，而且这种改动肯定应该是作为单独的 commit 。

不过目前看这部份代码算不上必须的。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1147,144 @@ def _restore_metadata(self, backup, volume_id):
+
+        return tmpdir
+
+    def _get_snapshot(self,backup):
```
_get_snapshot 里部份代码已经包含在 _rbd_image_exists 里。

通过已有的 _get_backup_base_name 和 _rbd_image_exists 就可以找到 backup 具体对应的 rbd 或者 snapshot 。

wwba replied:
原来调入了_get_restore_point,是有点问题.

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1147,144 @@ def _restore_metadata(self, backup, volume_id):
+                                   backup['id'])
+
+
+    def _full_upload(self, backup, image_service, image_meta, volume_name):
```
_full_upload 和下面 copy_backup_to_image 这个名字不太好。

其实从 cinder/api 那里开始，可以这样： 
```
cinder/api  -> cinder/backup/api -> cinder/backup/rpcapi        ->  cinder/backup/manager   ->
upload      -> upload            ->  upload_backup_to_image     -> upload_backup_to_image   -> 

cinder/backup/driver      -> cinder/backup/drivers/ceph
upload_backup_to_image    -> upload_backup_to_image    -> clone_backup_to_image
                                                       -> copy_backup_to_image
```

zhaochao commented on cinder/api/contrib/backups.py:
```diff
@@ -267,6 +270,74 @@ def create(self, req, body):
+        backup = self.backup_api.get(context, id)
+        try:
+            response = self.backup_api.copy_backup_to_image(context,
+                                                  backup,
```
这几行缩进有问题。

zhaochao commented on cinder/api/contrib/backups.py:
```diff
@@ -379,7 +450,7 @@ def get_resources(self):
             Backups.alias, BackupsController(),
             collection_actions={'detail': 'GET', 'import_record': 'POST'},
-            member_actions={'restore': 'POST', 'export_record': 'GET',
+            member_actions={'restore': 'POST', 'upload': 'POST', 'export_record': 'GET',
```
这一行太长了，需要换行。

风格问题，用 pep8  和 pylint 检查一下。

zhaochao commented on cinder/api/views/backups.py:
```diff
@@ -57,6 +57,28 @@ def restore_summary(self, request, restore):
 
+    def upload_summary(self, request, upload):
+        """Generic, non-detailed view of a restore."""
+        retval = {
```
这个 dict 的缩进太多

zhaochao replied:
这里缩进的问题还没改？（参见 pylint ）。

zhaochao commented on cinder/api/views/backups.py:
```diff
@@ -57,6 +57,28 @@ def restore_summary(self, request, restore):
+            'upload': retval
+        }
+
+
```
多了一个空行。

zhaochao commented on cinder/backup/api.py:
```diff
@@ -34,6 +34,10 @@
 from cinder.volume import utils as volume_utils
 
+from cinder.image import glance
+from cinder.openstack.common import excutils
```
前面已经 import 过了。

zhaochao commented on cinder/backup/api.py:
```diff
@@ -34,6 +34,10 @@
 
+from cinder.image import glance
+from cinder.openstack.common import excutils
+from cinder.image import image_utils
```
image_utils import 之后没用到？

zhaochao commented on cinder/backup/api.py:
```diff
@@ -54,6 +58,7 @@ class API(base.Base):
     def __init__(self, db_driver=None):
         self.backup_rpcapi = backup_rpcapi.BackupAPI()
         self.volume_api = cinder.volume.API()
+        self.image_service =  glance.get_default_image_service()
```
= 之后，多了 1 个空格。

zhaochao commented on cinder/backup/api.py:
```diff
@@ -202,6 +207,60 @@ def _consumed(resource_name):
+        try:
+            metadata['version'] = 1
+            recv_metadata = self.image_service.create(context,
+                         metadata)
```
缩进。

zhaochao commented on cinder/backup/api.py:
```diff
@@ -202,6 +207,60 @@ def _consumed(resource_name):
+            with excutils.save_and_reraise_exception():
+                pass
+
+
```
2 行空行太多。（pep8）

zhaochao commented on cinder/backup/api.py:
```diff
@@ -202,6 +207,60 @@ def _consumed(resource_name):
+        if 'is_public' in recv_metadata:
+            response['is_public'] = recv_metadata.get('is_public')
+        LOG.info(_("backup %(backup_id)s to image %(image_id)s completed successfully."),
+                 {'backup_id':backup['id'],'image_id':recv_metadata['id']})
```
分号、逗号之后需要 1 个空格（pep8）

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -61,6 +61,12 @@
+from cinder.openstack.common import fileutils
+from cinder.image import image_utils
+from cinder.openstack.common import processutils
+import urllib
```
urllib 没用到

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1147,144 @@ def _restore_metadata(self, backup, volume_id):
+        raise self.rbd.ImageNotFound(_("backup %s image not found") %
+                                   backup['id'])
+
+
```
只需要 1 行空行（pep8）

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1147,144 @@ def _restore_metadata(self, backup, volume_id):
+                    '--pool', self._ceph_backup_pool,
+                    '--cluster', strutils.safe_encode(CONF.backup_cluster_name),
+                    volume_name, tmp_file]
+            #args.extend(self._ceph_args())
```
注释的代码要去掉。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1147,144 @@ def _restore_metadata(self, backup, volume_id):
+                    volume_name, tmp_file]
+            #args.extend(self._ceph_args())
+            self._try_execute(*args)
+            image_utils.upload_volume(self.context, image_service,
```
upload_volume 要用的话可能需要重构一下，不然名称容易引起误解。

wwba replied:
这里上传的是一个临时文件, upload_volume的确不是很准确

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1147,144 @@ def _restore_metadata(self, backup, volume_id):
+                  {'src': backup['id']})
+
+        try:
+                backup_volume_name, snap = self._get_snapshot(backup)
```
缩进。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1147,144 @@ def _restore_metadata(self, backup, volume_id):
+
+        try:
+                backup_volume_name, snap = self._get_snapshot(backup)
+        except:
```
要捕获所有异常用 `except Exception` （虽然也不太好）

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1147,144 @@ def _restore_metadata(self, backup, volume_id):
+        do_clone_upload = False
+        if (image_meta['disk_format'] == 'raw' and
+                image_meta['container_format'] == 'bare'):
+           backend_upload = ceph_directclone.RbdBackupBackendUpload(self.context, backup,
```
从这一行开始基本都有缩进和空格的问题，可以对照 pep8 和 pylint 的检查看。

zhaochao commented on cinder/backup/drivers/clone/directclone.py:
```diff
@@ -0,0 +1,359 @@
+# Copyright 2013 Canonical Ltd.
```
这个文件的组织方式不太好。

1、关于 rbd 本身的操作，在 cinder/backup/drivers/ceph 中已经包含了基础代码，不需要重复；新增加的操作应该也直接加进去；
2、Glance Image 的操作则应该加到  cinder/image 中，比如 _get_image_ref 。

wwba replied:
_get_image_ref 是获取当前volume数据库中记录的image_id,应该放在RBDDriver,CephBackupDriver里面,而且有个问题: _get_image_ref 会使用cinder.volume, 放在image/image_utils.py然后导入cinder.volume显然不合适,因为image_utils.py 导入只是一些基本工具.

zhaochao replied:
噢，那这个 `_get_image_ref` 是我看错了，还以为是要从 glance api 获取信息。

不过应该不需要用到 volume_api ，`BackupDriver` 也有 `get_metadata` ，应该能取到所有 metadata 。

zhaochao commented on cinder/backup/drivers/clone/directclone.py:
```diff
@@ -0,0 +1,359 @@
+import fcntl
+import os
+import re
+import time
```
上面这几个模块没用到，应该去掉。

zhaochao commented on cinder/backup/drivers/clone/directclone.py:
```diff
@@ -0,0 +1,359 @@
+
+LOG = logging.getLogger(__name__)
+
+class RbdBackupBackendUpload(object):
```
类的定义与其它代码之间空 2 行（pep8）

zhaochao commented on cinder/backup/drivers/clone/directclone.py:
```diff
@@ -0,0 +1,359 @@
+        try:
+            # TODO: The easy way -- the image is an RBD clone, so use the parent
+            # images' storage pool
+            # parent_pool, _im, _snap = self.driver.parent_info(self.rbd_name)
```
全是注释？

zhaochao commented on cinder/backup/drivers/clone/directclone.py:
```diff
@@ -0,0 +1,359 @@
+        base_image_id =  self._get_image_ref()
+        fsid = self._get_fsid()
+
+        # is configured, but we can infer what storage pool Glance is using
```
注释不全？

zhaochao commented on cinder/backup/drivers/clone/directclone.py:
```diff
@@ -0,0 +1,359 @@
+        return metadata
+
+    def clone_backup_and_add_location(self, snapshot, 
+                             is_snapshot_exists, snap_protected = False):
```
is_snapshot_exists 这个参数有点多余了， 前一个参数 snapshot 已经包含所有需要的信息。

zhaochao commented on cinder/backup/drivers/clone/directclone.py:
```diff
@@ -0,0 +1,359 @@
+    def _supports_layering(self):
+        return hasattr(self.rbd, 'RBD_FEATURE_LAYERING')
+        
+    def _create_snapshot(self, snapshot, pool=None, protected=True):
```
`_create_snapshot`  、 `_protect_snapshot`  、 `_is_protect_snapshot` 、 `_unprotect_snapshot` 、`_flatten` 这几个函数的必要性需要讨论，在 `CephBackupDriver` 中 `rados` 和 `rbd` 都已经 import 进来，所以应该充分利用这两个模块，没有特别的原因一定要用 `RBDVolumeProxy` ，而且使用 Volume 字眼容易产生混淆。

zhaochao commented on cinder/backup/drivers/clone/directclone.py:
```diff
@@ -0,0 +1,359 @@
+
+
+
+    def _delete_cloned_image(self, name, dest_pool):
```
_delete_cloned_image 的作用过于宽泛了，这里不需要一个通用的删除 rbd 的方法，只需要清理在 clone backup to glance image 创建的 rbd 就行，不需要关注存不存在 snapshot （因为这个 rbd 是刚刚创建，不应该存在 snapshot ）。

另外，这里（以及其它地方）也一样，尽量把 volume 、backup 、image （ rbd 或者 Glance ）区分开。

wwba replied:
_delete_cloned_glance_image? 这里要删除快照的, location update失败的话, 这时候glance image 的名为'snap'的快照已经创建出来了.

zhaochao replied:
我的意思是，这个函数实现过于宽泛了。

比如删 snapshot ，`_delete_cloned_image` 就用了 2 次，一次是 snapshot 还没有创建，一次是 snapshot 已经创建，而且创建snapshot 也是这个流程中的代码创建，没有必要做 list_snaps 。

比如使用 RBDVolumeProxy（几个相关的函数都是这样），每次调用这些函数的时候传递是 snapshot 和 pool ，但是调用这些函数的地方经常都是几个连着使用，这样的话，就没有特别的必要每次都重新实例化 RBDVolumeProxy ，都用同一个 rbd.Image 对象就可以了。

wwba replied:
封裝RBDVolumeProxy可讀性要強一些,  _delete_cloned_image 內部是用一個rbd.Image來完成操作的.

zhaochao replied:
用同一个 rbd.Image 对象说的是 `_create_snapshot`  、 `_protect_snapshot`  、 `_is_protect_snapshot` 、 `_unprotect_snapshot` 、`_flatten`  这几个函数， 它们基本都只在`clone_backup_and_location` 里被调用，这样的话用同一个 `rbd.Image` ，都没有单独封装这几个函数。

`_delete_cloned_image` 的问题，前面已经说了，这个函数用法很固定。同时 `snapshot` 也是在 `clone_backup_and_location` 执行过程创建的，所以不需要做过多的检查。`snapshot` 已经创建的话，就先删 `snapshot` ，再删 `rbd` ；`snapshot` 还没创建，就直接删 `rbd` 。`snapshot` 有的话只会有 1 个 ，而且名字是确定，这样也不用 `list_snaps` 。

zhaochao commented on cinder/backup/drivers/clone/directclone.py:
```diff
@@ -0,0 +1,359 @@
+                         snap=snapshot['name'])}          
+        try:
+            self._clone(location, self.image_id, dest_pool=parent_pool)
+        except:
```
单独用 `except:` 的问题，见之前的评论。

zhaochao commented on cinder/backup/drivers/clone/directclone.py:
```diff
@@ -0,0 +1,359 @@
+        metadata = {
+                    'status': 'active',
+                    'name': str(self.image_meta['name']),
+                    #'properties': {
```
没有必要的注释应该删掉。

zhaochao commented on cinder/backup/drivers/clone/directclone.py:
```diff
@@ -0,0 +1,359 @@
+        image_snapshot = {}
+        image_snapshot['volume_name'] = self.image_id
+        image_snapshot['name'] = 'snap'
+        self._create_snapshot(image_snapshot, pool=parent_pool, protected=True)
```
`protected` 默认是 `True` ，这里不用重复。

zhaochao commented on cinder/backup/drivers/clone/directclone.py:
```diff
@@ -0,0 +1,359 @@
+        image_snapshot['name'] = 'snap'
+        self._create_snapshot(image_snapshot, pool=parent_pool, protected=True)
+        
+        metadata = self._create_snapshot_metadata()
```
感觉没必要单独用一个函数，而且命令也不太好，直接用类似于 updated_metadata ，然后直接赋值就可以了。

zhaochao commented on cinder/backup/drivers/clone/directclone.py:
```diff
@@ -0,0 +1,359 @@
+            self.image_service.update(self.context, str(self.image_id), metadata,
+                                 data=None, purge_props=False)
+          
+        except:
```
`except:` 的问题。

zhaochao commented on cinder/image/glance.py:
```diff
@@ -153,8 +153,9 @@ def call(self, context, method, *args, **kwargs):
-        if version in kwargs:
+        if 'version' in kwargs:
             version = kwargs['version']
+            kwargs.pop('version',None)
```
~~这个改动的原因？~~

~~如果是必须的改动，作为单独的一个 commit ，用 commit message 说明原因才行。~~

上游有相关的改动，应该直接 backport 上游的代码。

zhaochao reviewed (changes requested):
主要功能实现应该没有问题，但是：
1、代码组织不好，很多代码和已有代码重复，或者没有充分利用已有代码；
2、上游已有的改动需要 backport ，而不是直接修改；
3、代码风格，尽可能地通过 pep8 、pylint 的检查。

wwba committed:
Fix glance image create fail with glance api v1

Changes in _translate_from_glance() from commit
c67761b29f088ff3c2f2407e3a9617f80e98ecbf bring a error

Signed-off-by: Xiaojun Liao <xiaojunliao85@gmail.com>

zhaochao commented on cinder/api/contrib/backups.py:
```diff
@@ -267,6 +269,61 @@ def create(self, req, body):
+            raise webob.exc.HTTPBadRequest(explanation=msg)
+
+
+        image_metadata = {"container_format": params.get(
```
前面 2 行空行太多（pep8）；
后面几行缩进不太行（pylint）。

zhaochao commented on cinder/backup/api.py:
```diff
@@ -202,6 +204,61 @@ def _consumed(resource_name):
+
+        try:
+            image_service = glance.get_default_image_service()
+            metadata['version'] = 1
```
看 cinder/image/glance 的实现， 把 `version` 放到 metadata 里，感觉不是很好的做法，只能算是一个 trick 。

创建 image 的时候，不需要用到 glance api v1 。

另外这里需要设置 image 的 `min_disk` ，应该等于 backup 的大小。

其它相关于 glance api v2 的说明会在 update location 和 cinder/image/glance 里说明。

zhaochao commented on cinder/api/contrib/backups.py:
```diff
@@ -267,6 +269,61 @@ def create(self, req, body):
+            msg = _("No image_name was specified in request.")
+            raise webob.exc.HTTPBadRequest(explanation=msg)
+
+        is_public = params.get('is_public', 'False')
```
我重新看了下 volume upload 以及 nova snapshot 的代码，从其它组件创建 image 时应该不需要支持更多的参数，像 `is_public` 、`protected` 就不需要了，这些属性正常应该直接通过 glance api 来完成，不在 cinder 这里给用户过多的选择。

如果一定要支持这些参数的话，也会影响到后面关于 cinder 里关于 glance api 的调用，因为像 `is_public` ，在 glance api v1 和 v2 里属性名称已经发生了变化。

wwba replied:
is_public glance api v2的确不能使用.

zhaochao replied:
用的话是 `visibility` ，不过还是不要在 cinder 里设置 `is_public` 和 `protected` 这些参数了。

zhaochao commented on cinder/api/views/backups.py:
```diff
@@ -57,6 +57,27 @@ def restore_summary(self, request, restore):
+                  'disk_format': upload['disk_format'],
+                  'image_name': upload['image_name']
+                 }
+        if 'protected' in upload:
```
`protected` 和 `is_public` 参见前面的评论。

zhaochao commented on cinder/backup/api.py:
```diff
@@ -202,6 +204,61 @@ def _consumed(resource_name):
+                    "container_format": recv_metadata['container_format'],
+                    "disk_format": recv_metadata['disk_format'],
+                    "image_name": recv_metadata.get('name', None)}
+        if 'protected' in recv_metadata:
```
`proteced` 和 `is_public` 见前面的评论。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -61,6 +61,14 @@
 
+from cinder.openstack.common import fileutils
+from cinder.image import image_utils
+from cinder.openstack.common import processutils
```
processutils 没用到？

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
             raise exception.BackupOperationError(msg)
 
+    def _get_fsid(self):
+        with rbd_driver.RADOSClient(self, self._ceph_backup_pool) as client:
```
RADOSClient 在整个 upload backup to image 的过程中应该只需要实例化 2 次：打开 backup pool 和 image pool ，其关所有相关操作因为都使用这 2 个实例完成。

zhaochao replied:
后面 clone_backup_to_image 里已经有 with rbd_driver.RadosClient 了，get_fsid 可以直接在 with 语句块之类完成。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
+        with rbd_driver.RADOSClient(self, self._ceph_backup_pool) as client:
+            return client.cluster.get_fsid()
+
+    def _parse_url(self, url):
```
函数和变量命名上用 _parse_image_loc 、 location 之类好点。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
+        # Find the first location that is in the same RBD cluster
+        for location in locations:
+            if location is None:
+                contine
```
typo. 
contine -> continue

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
+        parent_pool = None
+
+        try:
+            image_service = glance.get_default_image_service()
```
image_service 在 cinder/backup/manager 已经创建了一个实例，也传递给 cinder/backup/drivers/ceph 了，用这一个实例就可以了。

zhaochao replied:
这里没改？``upload_backup_to_image`` 接收了一个 image service 的实例，后续的函数可以直接用。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
+            raise exception.ImageUnacceptable(
+                _('Cannot determine the parent storage pool for %s; '
+                  'cannot determine where to store images') %
+                base_image_id)
```
看 _get_parent_pool 被调用的地方， 应该没有必要 raise 异常？

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
+
+        return parent_pool
+
+    def _create_snapshot(self, snapshot, pool=None, protected=True):
```
`_create_snapshot` 、`_protect_snapsht` 、 `_is_protect_snapshot` 、`_delete_snapshot` 、`_delete_cloned_glance_image` 、`_clone` 、 `_flatten` 这几个函数，如果要单独作为函数的话，不应该总是把 pool 、rbd 和 snapshot 的名字传递进去，而应该直接把 RADOSClient 和 rbd.Image 实例传递，这样就不需要每次重新建立 rados 连接或者打开 rbd 了。

直接使用 RADOSClient 和 rbd.Image 实例之后，很多代码都可以简化。

wwba replied:
封装snapshot相关的接口的时候,我们是不用考虑这些接口在上层调用函数被调用的次数,考虑实现完整的封装即可,这些接口可以参考RBDDriver的create_snapshot,delete_snapshot等,如果如你所说采用在上层调用RBDVolumeProxy()接口,然后将返回传给_create_snapshot(),上面你提到的这几个函数就没有作为CephBackupDriver方法的必要,而是直接在clone_backup_to_image()中实现了.但是,这样真的好吗?  性能考虑? 函数的封装都会有这样的问题,好像这里不是我们应该考虑的问题.

zhaochao replied:
对，我之前说过，目前看是没有单独封装的必要，如果要封装的话，我也觉得直接传递 rbd.Image 比较好，完全没必要每次都重新实例化 RBDVolumeProxy 。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
+        """Creates an rbd snapshot."""
+        with rbd_driver.RBDVolumeProxy(self, snapshot['volume_name'],
+                                       pool) as volume:
+            snap = strutils.safe_encode(snapshot['name'])
```
感觉 strutils.safe_encode 用得太多，对于 rbd 和 snap 这些可以在调这些函数之前做一次 safe_encode 。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
+                                        image=image_meta['id']))
+
+        try:
+            image_service = glance.get_default_image_service()
```
这里和 `_get_parent_pool` 里一样，应该直接使用 cinder/backup/manager 传递进来的 `image_service` 。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
+        metadata['version'] = 1
+        metadata['location'] = str('rbd://%(fsid)s/%(pool)s/%(image)s/snap' %
+                                   dict(fsid=fsid, pool=parent_pool,
+                                        image=image_meta['id']))
```
继续说明关于 glance api v2 update image 的问题：

我花了一些时间研究 glance api v2 以及 glanceclient ，确定通过 v2 也是可以的，只不过 api 请求内容有变化：
https://developer.openstack.org/api-ref/image/v2/?expanded=update-an-image-detail#update-an-image

另外， update 是应该只需要更新 `status` 和 `location` 就可以，别的不应该改动。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
+
+    def _image_conversion_dir(self):
+        tmpdir = (CONF.image_conversion_dir or
+                  tempfile.gettempdir())
```
tempfile 没有 import ？

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
+        return tmpdir
+
+    def _get_snapshot(self, backup):
+        """returns: [volume_name,snap]"""
```
只 return 确定2个元素的话，用 tuple ？ 

volume_name -> rbd_name ？

wwba replied:
ceph的驱动,用rbd_name更合适.

wwba replied:
跟这个无关的, 发现copy_backup_to_image()没有考虑snap的情况,有点问题.

zhaochao replied:
噢，对，`copy_backup_to_image` 这个看漏了

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
+            raise exception.InvalidParameterValue(msg)
+
+        with rbd_driver.RADOSClient(self, self._ceph_backup_pool) as client:
+            # diff backup
```
cinder/backup/drivers/ceph 里 rbd 命名其实是 2 种 format ： diff format 和 old style format 。

这个和 diff backup 和 full backup 没有关系，现在即使是 full backup 也是 diff format 。old style format 我们这里没有，不过代码是要支持。所以注释里可以改下。

wwba replied:
old stype format 比较模糊, 改no snap format ?

zhaochao replied:
看原有的代码注释都是这么叫的，直接用就可以了，换个名字别人不一定明白。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
+        """Copy backup to a local file"""
+        with fileutils.file_open(tmp_path, 'wb') as dest_file:
+            try:
+                with rbd_driver.RADOSClient(self,
```
这里也一样，应该尽可能复用 RADOSClient 和 rbd.Image 。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
+
+        with fileutils.remove_path_on_error(tmp_path):
+            self._copy_backup_to_file(volume_name, tmp_path, length)
+            image_service = glance.get_default_image_service()
```
image_sesrvice 可以复用。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,415 @@ def _restore_metadata(self, backup, volume_id):
+                                    str(dest_name),
+                                    features=self.rbd.RBD_FEATURE_LAYERING)
+                except rbd.PermissionError:
+                    raise exception.Forbidden(_('no write permission on '
```
exception.Forbidden 没定义？

zhaochao commented on cinder/image/glance.py:
```diff
@@ -335,48 +332,20 @@ def delete(self, context, image_id):
             raise exception.ImageNotFound(image_id=image_id)
         return True
 
-    def _translate_from_glance(self, context, image):
```
这里我是倾向于直接使用 glance api v2 ，但是需要进行以下改动：
1、传递进来的 metadata ，必须符合 glance api v2 要求，当然衍生的新问题是，需要针对 glance api v1 和 glance api v2 单独构造 update 时的 metadata ；
2、glanceclient 里有一个 commit 需要 backport ：
https://git.openstack.org/cgit/openstack/python-glanceclient/commit/?id=90407d9e473014c24eeab294192f9d3208f58ea7

zhaochao commented on cinder/backup/rpcapi.py:
```diff
@@ -57,6 +57,12 @@ def restore_backup(self, ctxt, host, backup_id, volume_id):
                    volume_id=volume_id)
 
+    def upload_backup_to_image(self, ctxt, host, backup, image_meta):
+        LOG.debug("copy_backup_to_image in rpcapi backup_id %s", backup['id'])
```
这里 debug 日志也忘记改了， copy_backup_to_image -> upload_backup_to_image 。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,424 @@ def _restore_metadata(self, backup, volume_id):
+        """Clone volume and register its location to the image."""
+
+        base_image_id = self._get_image_ref(backup['volume_id'])
+        fsid = self._get_fsid()
```
按上面说的 _get_fsid 完全可以放到 with RadosClient() context 里面去。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,424 @@ def _restore_metadata(self, backup, volume_id):
+        if parent_pool is None:
+            return False
+
+        is_snap_exists = bool(snapshot['name'])
```
这里没有必要专门增加一个变量了，后面针对 is_snapshot_exists 的判断，都可以直接把 is_snapshot_exits 替换成snapshot['name'] ，不会有问题，也不影响理解代码。

wwba replied:
当时这样考虑是因为觉得加一个变量可读性要强一些．

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,424 @@ def _restore_metadata(self, backup, volume_id):
+        volume_name = snapshot['volume_name']
+        snap = snapshot['name']
+        if not is_snap_exists:
+            snapshot['name'] = uuid.uuid4().hex
```
上面 3 行有点问题。看 clone_backup_to_image 后面是没有用到 snapshot 这变量的，反而是如果 clone_backup_to_image 失败了，snapshot 这个变量还有机会被 copy_backup_to_image 用到。

反而是 snap 这个变量一直是用的 snapshot 里原来 name 这个 key 的值，也就是如果 snapshot['name'] 本来是空的话， snap 一直会是 None ，而 snapshot['name'] 会使用一个新的 uuid ，但是一直到 copy_backup_to_image 才用上。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,424 @@ def _restore_metadata(self, backup, volume_id):
+                                    str(self._ceph_backup_pool)) as src_client:
+            try:
+                volume = self.rbd.Image(src_client.ioctx,
+                                        strutils.safe_encode(volume_name),
```
这里包括后面的 safe_encode 都可以在前面给 volume_name 以及 snap 赋值的时候处理，可以避免不断的使用 safe_encode。（snap 这个变量还有另外一个问题，参见 L1249 的评论）。

wwba replied:
这个问题我也考虑过，按照cinder其他地方对rbd api调用来看，用些地方是不需要safe_encode的，所以volume_name和snap没有经过safe_encode的处理．

zhaochao replied:
这应该不是什么问题， safe_encode 只是用 utf-8 重新编码，不会影响别的接口使用。

wwba replied:
日志打印不需要safe_encode,　使用safe_decode?

zhaochao replied:
safe_encode 对日志输出应该没什么影响。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,424 @@ def _restore_metadata(self, backup, volume_id):
+                    strutils.safe_encode('snap'))
+            except self.rbd.ImageBusy:
+                raise exception.SnapshotIsBusy(
+                    snapshot_name='snap')
```
上面的 if 判断一直到这里，再改动的代码里出现了不少次，这里可以重新设计一下。

self._supports_layering 是 clone_backup_to_image 的基本条件，所以应该在 upload_backup_to_image 里开始 clone_backup_to_image 之前就进行判断，如果不支持 layering ，直接进入 copy_backup_to_image 的流程。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,424 @@ def _restore_metadata(self, backup, volume_id):
+                    try:
+                        volume.close()
+                    finally:
+                        dest_volume.close()
```
这里是在进行初始化 dest_volume 的异常处理，也就是 dest_volume 应该都没有初始化完成，也就不用 close 了。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,424 @@ def _restore_metadata(self, backup, volume_id):
+                try:
+                    volume.close()
+                finally:
+                    dest_volume.close()
```
上面 dest_volume 几步操作 flatten 、create_snap 、protect_snap 以及 image update 的异常处理都是相同的，可以合并，放到 finally 里。

zhaochao replied:
另外一种方案是还是用回去 RBDVolumeProxy ，不过和 rados.RADOSClient 一样只需要 2 个实例就可以了。使用 RBDVolumeProxy 的好处是上面 volume.close() 和 dest_volume.close() 都可以省略了，当然 _delete_cloned_glance_image 还是一样，统一放到 finally 之类的地方。

噢，rbd.RBD().clone 那里是要有 1 个 rados.RADOSClient 的实例，这样会稍微麻烦一点，但是也还可以接收。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,424 @@ def _restore_metadata(self, backup, volume_id):
+                                   dest_name=image_meta['id']))
+                    self.rbd.RBD().clone(src_client.ioctx,
+                                         volume_name.encode('utf-8'),
+                                         snap.encode('utf-8'),
```
这里 encode 也是一样，一是统一用 strutils.safe_encode ， 二是尽量一开始就 encode ，不要到用时才处理。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,424 @@ def _restore_metadata(self, backup, volume_id):
+            msg = _("Backup id required")
+            raise exception.InvalidParameterValue(msg)
+
+        with rbd_driver.RADOSClient(self, self._ceph_backup_pool) as client:
```
_get_snapshot 在 upload_snapshot 一开始就调用了，也就是说 backup 对应的 rbd 应该不开始就通过 RADOSClient 或者 RBDVolumeProxy 打开，而这个 RADOSClient 或者可以一直用到整个功能结束（不论是 clone_backup_to_image 还是 copy_backup_to_image 都可以直接使用）。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,424 @@ def _restore_metadata(self, backup, volume_id):
+
+    def _copy_backup_to_file(self, snapshot, tmp_path, length):
+        """Copy backup to a local file"""
+        rbd_name = strutils.safe_encode(snapshot['volume_name'])
```
这个函数里的 strutils.safe_encode 而前面是同样的情况，可以以参数形式传进来之前先行 encode ）。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,424 @@ def _restore_metadata(self, backup, volume_id):
+        with fileutils.file_open(tmp_path, 'wb') as dest_file:
+            try:
+                with rbd_driver.RADOSClient(self,
+                                            self._ceph_backup_pool) as client:
```
这里看起来可能直接用 RBDVolumeProxy ，如果调整之后，在 upload_backup_to_image 里面已经实例化一个 RBDVollumeProxy 的话，这里可以直接使用这个实例。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,424 @@ def _restore_metadata(self, backup, volume_id):
+        LOG.debug('Starting upload from Ceph backup=%(src)s to glance',
+                  {'src': backup['id']})
+
+        try:
```
按照前面提到的，可以在 _get_snapshto 之前，先对 backup 实例化一个  RBDVolumeProxy ，这个实例可以一直使用到整个功能结束。

wwba replied:
像之前我们讨论的RBDVolumeProxy，RADOSClient做为参数传入被调用函数问题，会增加调用与被调用调用函数的耦合性(函数参数会增加)，调用RBDVolumeProxy的_copy_backup_to_file是获取的快照状态与其他两个函数还不一样．

zhaochao replied:
嗯，_get_snapshot 这里倒是可以不改动。

但是 _get_snapshot 之后 backup 对应是一个 rbd 还是一个 snapshot 已经确定，可以直接初始化一个 RBDVolumeProxy 实例，这样不论是 clone_backup_to_imge 还是 copy_backup_to_image 都可以直接使用。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,424 @@ def _restore_metadata(self, backup, volume_id):
+
+        snapshot = {}
+        snapshot['volume_name'] = backup_volume_name
+        snapshot['name'] = snap
```
前面的 comment 也提到过了，像 snap 这种，可以提前到这里进行 encode 。

zhaochao commented on cinder/backup/drivers/ceph.py:
```diff
@@ -1135,6 +1146,424 @@ def _restore_metadata(self, backup, volume_id):
+        do_clone_upload = False
+        if (image_meta['disk_format'] == 'raw' and
+                image_meta['container_format'] == 'bare'):
+            if CONF.image_upload_use_clone:
```
前面的 comment 提到过，可以在这里同时判断 self._supports_layering 。
